### PR TITLE
Bug in loc not ordering rhs correctly for mixed indexer

### LINKED ISSUE #40480: BUG: setting values with array in mixed DataFrame disregards order of indexer
```
In [2]: df = DataFrame(index=[3, 5, 4], columns=["A", "B"], dtype=float)
   ...: df["B"] = "string"
   ...: df.loc[[4, 3, 5], "A"] = np.array([1, 2, 3], dtype="int64")

In [3]: df
Out[3]: 
   A       B
3  1  string
5  2  string
4  3  string
```

In the above (using master), the values [1, 2, 3] were set in that order, while the indexer used a different order. In released versions of pandas, we correctly get [2, 3, 1]:

```
In [3]: df
Out[3]: 
   A       B
3  2  string
5  3  string
4  1  string

In [4]: pd.__version__
Out[4]: '1.2.3'

```

### PR DESCRIPTION
**Pull request by phofl**
- [x] closes #40480
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them

### CODE REVIEW

**Comment by jreback** [comment]
looks fine. is this a user facing change or only a patch on AM?

also merge master

**Comment by phofl** [comment]
This is a regression on master, so not user facing

**Comment by jreback** [comment]
is the array manager failure related?

**Comment by phofl** [comment]
Sorry, did not scroll down far enough the last time.

This fixes part of the array manager bug, but not the overwriting, so adjusted the test accordingly

**Comment by jreback** [comment]
great thanks @phofl

File: pandas/tests/indexing/test_loc.py [hunk 1]
```diff
@@ -616,8 +612,16 @@ def test_loc_setitem_frame_with_reindex_mixed(self):
         expected["B"] = "string"
         tm.assert_frame_equal(df, expected)
 
+    def test_loc_setitem_frame_with_inverted_slice(self):
+        # GH#40480
+        df = DataFrame(index=[1, 2, 3], columns=["A", "B"], dtype=float)
+        df["B"] = "string"
+        df.loc[slice(3, 0, -1), "A"] = np.array([1, 2, 3], dtype="int64")
+        expected = DataFrame({"A": [3, 2, 1], "B": "string"}, index=[1, 2, 3])
```
Comment (jbrockmendel) [inline pandas/tests/indexing/test_loc.py:620]: @phofl AFAICT the setting on L619 should be done in-place, so `expected["A"]` should still be float dtype.  is the changed dtype relevant here?

Comment (jbrockmendel) [reply pandas/tests/indexing/test_loc.py:620]: ditto for test_loc_setitem_frame_with_reindex_mixed

**Pull request merged**